Attach 'send' autograd function to the autograd graph as part of RPC.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24876 Attach 'send' autograd function to the autograd graph as part of RPC.**

This contains very basic functionality of adding 'send' autograd
function to our autograd graph. The purpose of this change is to validate the
basic structure proposed here makes sense. Once this makes sense, we can build
upon this to address more complicated scenarios. At a high level we've added
the following functionality:

1) Define a very simple 'SendRpcBackwards' autograd function.
2) Attach this function to appropriate tensors when we call an RPC.
3) Store the send function in our distributed autograd context.

GitHub Issue: https://github.com/pytorch/pytorch/issues/23110
Differential Revision: [D16903255](https://our.internmc.facebook.com/intern/diff/D16903255/)

Commit:
Attach 'send' autograd function to the autograd graph as part of RPC.

This contains very basic functionality of adding 'send' autograd
function to our autograd graph. The purpose of this change is to validate the
basic structure proposed here makes sense. Once this makes sense, we can build
upon this to address more complicated scenarios. At a high level we've added
the following functionality:

1) Define a very simple 'SendRpcBackwards' autograd function.
2) Attach this function to appropriate tensors when we call an RPC.
3) Store the send function in our distributed autograd context.

Differential Revision: [D16903255](https://our.internmc.facebook.com/intern/diff/D16903255/)

mrshenli commented on tools/build_variables.py:
```diff
@@ -49,8 +49,10 @@
+    "torch/csrc/distributed/autograd/Utils.cpp",
     "torch/csrc/distributed/autograd/context/DistAutogradContainer.cpp",
     "torch/csrc/distributed/autograd/context/DistAutogradContext.cpp",
+    "torch/csrc/distributed/autograd/functions/SendRpcBackwards.cpp",
```
we are using snake names now

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -37,6 +41,20 @@ std::shared_ptr<FutureMessage> py_rpc_builtin(
+              torch::jit::last(stack, op->schema().arguments().size());
+
+          // Attach the appropriate autograd edges.
+          auto grad_fn = addSendRpcBackward(ivalues);
```
How does this work for `RRef` args?

pritamdamania87 replied:
We discussed this offline, as part of .to_here() we would attach the necessary args.

mrshenli commented on torch/csrc/distributed/autograd/functions/SendRpcBackwards.h:
```diff
@@ -0,0 +1,16 @@
+namespace distributed {
+namespace autograd {
+
+struct TORCH_API SendRpcBackwards : public torch::autograd::Node {
```
Should this be called `SendRpcBackward` instead? Not 100% sure, just want to make sure we are consistent with other backward node.

mrshenli commented on torch/csrc/distributed/autograd/functions/SendRpcBackwards.h:
```diff
@@ -0,0 +1,16 @@
+  torch::autograd::variable_list apply(
+      torch::autograd::variable_list&& grads) override;
+};
+
```
We would also need a `RecvRpcBackward` to pair with this `SendRpcBackward`, no?

pritamdamania87 replied:
Yes, I wanted to keep this PR simple and added only `send` for early feedback on the general design. Will follow up on `recv` in a different PR.

mrshenli commented on test/test_rpc.py:
```diff
@@ -223,6 +230,41 @@ def test_py_function_exception(self):
+            # Get send function.
+            ctx = dist_autograd._current_context()
+            self.assertEqual(context_id, ctx._context_id())
+            send_functions = ctx._send_functions()
```
Is this just for temporary test purpose, or do we plan to keep this API in the final version?

mrshenli commented on torch/csrc/distributed/autograd/context/DistAutogradContext.cpp:
```diff
@@ -12,6 +12,17 @@ int64_t DistAutogradContext::context_id() const {
+  sendAutogradFunctions_.push_back(func);
+}
+
+std::vector<std::shared_ptr<SendRpcBackwards>> DistAutogradContext::
```
Should this be a map, as recv need to use its ID to find the corresponding send??

pritamdamania87 replied:
We can modify this in follow up PRs to a map, right now we don't have an ID to put in this map.

Commit:
Update on "Attach 'send' autograd function to the autograd graph as part of RPC."


This contains very basic functionality of adding 'send' autograd
function to our autograd graph. The purpose of this change is to validate the
basic structure proposed here makes sense. Once this makes sense, we can build
upon this to address more complicated scenarios. At a high level we've added
the following functionality:

1) Define a very simple 'SendRpcBackwards' autograd function.
2) Attach this function to appropriate tensors when we call an RPC.
3) Store the send function in our distributed autograd context.

GitHub Issue: https://github.com/pytorch/pytorch/issues/23110
Differential Revision: [D16903255](https://our.internmc.facebook.com/intern/diff/D16903255/)

ezyang commented:
When you get a chance please rebase past master failures

aazzolini commented:
@pritamdamania87 , how are we going to do this when the object containing the tensor is a python object?

pritamdamania87 commented:
@aazzolini As we discussed offline, when we pickle python objects, we can hook into the pickler and whenever we see tensors, we can attach them to the `send` autograd function appropriately.

Commit:
Update on "Attach 'send' autograd function to the autograd graph as part of RPC."


This contains very basic functionality of adding 'send' autograd
function to our autograd graph. The purpose of this change is to validate the
basic structure proposed here makes sense. Once this makes sense, we can build
upon this to address more complicated scenarios. At a high level we've added
the following functionality:

1) Define a very simple 'SendRpcBackwards' autograd function.
2) Attach this function to appropriate tensors when we call an RPC.
3) Store the send function in our distributed autograd context.

GitHub Issue: https://github.com/pytorch/pytorch/issues/23110
Differential Revision: [D16903255](https://our.internmc.facebook.com/intern/diff/D16903255/)

ezyang commented on torch/csrc/distributed/autograd/functions/sendrpc_backwards.h:
```diff
@@ -0,0 +1,16 @@
+namespace distributed {
+namespace autograd {
+
+struct TORCH_API SendRpcBackwards : public torch::autograd::Node {
```
This gunna need a docblock

ezyang commented:
> The purpose of this change is to validate the basic structure proposed here makes sense.

How do we actually validate this?

ezyang commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -41,6 +45,20 @@ std::shared_ptr<FutureMessage> py_rpc_builtin(
         Stack stack = torch::jit::createStackForSchema(
```
This is not part of this PR, but I noticed it while reviewing the code here. Above, in line 38, there is a `opName.rfind("aten", 0) == 0`. That's a code smell to me: I generally don't think string matching in this way is a good idea, because it's easy to get confounded by cases like `"atenblah"`. Is there a reason we can't use `Symbol::fromQualString` and then test if the namespace is ATen? That seems semantically better to me.

ezyang commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -41,6 +45,20 @@ std::shared_ptr<FutureMessage> py_rpc_builtin(
+
         return agent.send(
             dstName, ScriptCall(op, std::move(stack)).toMessage());
       } catch (std::runtime_error) {
```
Also not part of this PR, but can we bound the try-catch block around the only scope which you actually expect to throw--the comment above implies `createStackForSchema` is the only thing that is currently throwing? This will help us from accidentally catching errors we don't intend to catch. It's very simple to do, while we're waiting for someone to actually refactor the downstream function in question. (Or you could just do that refactor?)

ezyang commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -41,6 +45,20 @@ std::shared_ptr<FutureMessage> py_rpc_builtin(
             op->schema(), args, kwargs, c10::nullopt);
 
+        auto& autogradContainer = DistAutogradContainer::getInstance();
+        if (autogradContainer.hasValidContext()) {
```
Hmm. So if there is no current valid context, we just silently don't attach any autograd edges. That doesn't sound right to me: if there is no valid context, we had better *error* if we try to backpropagate.

Basically, if I don't do a context, and then I attend to rpc and backward through it, what error message do I get? That's something you should be testing for.

ezyang commented on torch/csrc/distributed/autograd/utils.h:
```diff
@@ -0,0 +1,22 @@
+// and attaches the provided IValues as next_edges to the 'send' function. Only
+// IValues of type 'Tensor' are attached.
+// TODO: Support Tensors which might be nested inside other IValue types
+//       (ex: List, Tuple)
```
This is actually a serious limitation, because it means that operations which take tensor lists like `torch.stack` silently don't work with the current implementation right now. You don't have to make it work immediately, but we should really give a good error message if you try to use something bad.

ezyang commented on torch/csrc/distributed/autograd/context/dist_autograd_container.cpp:
```diff
@@ -38,16 +40,36 @@ const DistAutogradContext& DistAutogradContainer::newContext() {
+      next_context_id_ >
           (kMaxContextId |
            (static_cast<int64_t>(worker_id_) << kContextIdBits))) {
     throw std::runtime_error("We have run out of autograd context ids!!!");
```
nit: Not part of this PR, but we could have used a `TORCH_INTERNAL_ASSERT` here

ezyang replied:
(You don't have to do anything to this code, I'm just letting you know in case you need to write something similar in the future)

ezyang commented on torch/csrc/distributed/autograd/context/dist_autograd_container.cpp:
```diff
@@ -38,16 +40,36 @@ const DistAutogradContext& DistAutogradContainer::newContext() {
+DistAutogradContext& DistAutogradContainer::currentContext() {
+  TORCH_CHECK(
+      hasValidContext(),
+      "Current thread doesn't have a valid autograd context.");
```
So, if I think about what the classic mistake that a user will make, it is that they will write distributed RPC code without actually setting the context. Is this the error message that will occur if this happens? If so, we could give more clear guidance about how to fix one's code here, rather than just stating what the internal violated invariant is.

ezyang commented on torch/csrc/distributed/autograd/context/dist_autograd_container.cpp:
```diff
@@ -38,16 +40,36 @@ const DistAutogradContext& DistAutogradContainer::newContext() {
+      hasValidContext(),
+      "Current thread doesn't have a valid autograd context.");
+  std::lock_guard<std::mutex> guard(autograd_context_lock_);
+  if (autograd_context_.find(current_context_id_) == autograd_context_.end()) {
```
Nit: `find()` returns an iterator, which you can use to retrieve the element in question. The way you've written it here, you have to do *two* map lookups in the successful case when only one is necessary.

ezyang commented on torch/csrc/distributed/autograd/context/dist_autograd_container.cpp:
```diff
@@ -57,10 +79,15 @@ void DistAutogradContainer::releaseContext(int64_t context_id) {
+  if (current_context_id_ == context_id) {
+    // Reset the thread_local current context id, since it is no longer valid.
+    current_context_id_ = -1;
+  }
```
Not part of this PR per se, but do we have a C++-side RAII guard for allocating and releasing thread local context? It seems like a logical thing that should be part of the C++ API.

mrshenli replied:
C++ API is non-goal for now, but let's add a TODO here in case we forgot.

ezyang commented on torch/csrc/distributed/autograd/init.cpp:
```diff
@@ -28,23 +29,44 @@ PyObject* dist_autograd_init(PyObject* /* unused */) {
+          .def("_send_functions", [](const DistAutogradContext& ctx) {
+            std::vector<py::object> funcs;
+            for (const auto& sendFunction : ctx.sendFunctions()) {
+              funcs.push_back(py::reinterpret_borrow<py::object>(
```
I'm... pretty sure this is leaking Python objects. Inspection of `functionToPyObject` reveals that it incref objects before returning. `reinterpret_borrow` is the "copying" constructor: it doesn't take ownership of the input.

ezyang replied:
(If I'm right, you should use `reinterpret_steal` here instead.)

ezyang reviewed (changes requested):
Leaking Python objects, I think.

ezyang commented:
To summarize the overall review: this looks fine, but it isn't really enough to validate if the design works or not (you implemented X, you wrote a unit test that shows if X works, but there isn't enough of the actual guts to built out to know if X is what you actually want). That's probably fine: this patch is pretty localized, so we should let it in ASAP and keep working on the real functionality. I am blocking land because I think it is leaking Python objects as it stands.

Commit:
Update on "Attach 'send' autograd function to the autograd graph as part of RPC."


This contains very basic functionality of adding 'send' autograd
function to our autograd graph. The purpose of this change is to validate the
basic structure proposed here makes sense. Once this makes sense, we can build
upon this to address more complicated scenarios. At a high level we've added
the following functionality:

1) Define a very simple 'SendRpcBackwards' autograd function.
2) Attach this function to appropriate tensors when we call an RPC.
3) Store the send function in our distributed autograd context.

GitHub Issue: https://github.com/pytorch/pytorch/issues/23110
Differential Revision: [D16903255](https://our.internmc.facebook.com/intern/diff/D16903255/)

Commit:
Update on "Attach 'send' autograd function to the autograd graph as part of RPC."


This contains very basic functionality of adding 'send' autograd
function to our autograd graph. The purpose of this change is to validate the
basic structure proposed here makes sense. Once this makes sense, we can build
upon this to address more complicated scenarios. At a high level we've added
the following functionality:

1) Define a very simple 'SendRpcBackwards' autograd function.
2) Attach this function to appropriate tensors when we call an RPC.
3) Store the send function in our distributed autograd context.

GitHub Issue: https://github.com/pytorch/pytorch/issues/23110
Differential Revision: [D16903255](https://our.internmc.facebook.com/intern/diff/D16903255/)

pritamdamania87 commented:
> How do we actually validate this?

The PR was more to validate I was building the custom autograd function appropriately and attaching the tensors.

zhaojuanmao commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -1,14 +1,36 @@
+  // Record appropriate autograd information before sending the message over the
+  // wire.
+  auto& autogradContainer = DistAutogradContainer::getInstance();
+  if (autogradContainer.hasValidContext()) {
```
we may not be ready to attach C++ send/recv function for Pyper RPC integration yet.  But init_model_parallel() always init autograd and thus haveValidContext() is true. do we want to pass a disable_dist_auto_grad flag to  init_model_parallel() for now so that Pyper RPC integration can still use Python dist autograd version?

pritamdamania87 replied:
`init_model_parallel` doesn't make `hasValidContext()` true. The user needs to explicitly use `with torch.distributed.autograd.context()` around a block of code to enable any of the functionality here.

mrshenli commented on torch/csrc/distributed/autograd/context/dist_autograd_container.cpp:
```diff
@@ -57,10 +79,15 @@ void DistAutogradContainer::releaseContext(int64_t context_id) {
+
+  if (current_context_id_ == context_id) {
+    // Reset the thread_local current context id, since it is no longer valid.
+    current_context_id_ = -1;
```
Will this work for reentrant backward without nested context? (of course, reentrant backward support does not need to be part of the first release)

pritamdamania87 replied:
The user will probably invoke a specialized `torch.distributed.backward()` API. This will under the covers only have one autograd context. The engine itself can run multiple re-entrant backward passes under the hood, but we will associate all those backward passes with a single context.

mrshenli commented:
A code structure question: should `dist_autograd` be under the `rpc` folder? (if so, don't need to be addressed in this PR). I am asking because I was thinking about where we should put the `worker_id_t` type. This type should be shared by `rpc` and `dist_autograd`, and no other modules. But it does not make much sense to put this type directly under `distributed` folder. 

Do we foresee any use case for `dist_autograd` without using `rpc`? If not in near future, shall we merge `dist_autograd` into `rpc`?

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -35,36 +31,30 @@ std::shared_ptr<FutureMessage> py_rpc_builtin(
-          auto& currentContext = autogradContainer.currentContext();
-          currentContext.addSendFunction(grad_fn);
+        } catch (std::runtime_error& e) {
+          LOG(INFO) << "Couldn't match schema: " << op->schema()
```
This will print a lot of logs, I remember seeing ~20 overloaded versions of `aten::add`

mrshenli replied:
How many logs do you see when run tests locally?

pritamdamania87 replied:
I didn't push the latest version here, I've changed this to `VLOG(1)` so it will print only when verbosity is on.

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -1,14 +1,36 @@
+  if (autogradContainer.hasValidContext()) {
+    // Attach the appropriate autograd edges to the tensors found in the
+    // message.
+    auto grad_fn = addSendRpcBackward(message.tensors());
```
I am not against adding it here in this PR. But in the future, we might want to decouple `RpcAgent` interface with autograd code. @satgera mentioned that eventually, we might want to have a separate rpc lib so that `RpcAgent` implementations can have a minimum set of dependencies.

pritamdamania87 replied:
Yes, I have re-worked this a bit in my follow up PR for the full forward pass.

mrshenli commented:
Stamped to unblock, but please get another review from @ezyang and address @zhaojuanmao's comments before landing.

ezyang reviewed (approved):
It's too early to tell if this all works, but the diff is not too long so let's move it along.

pritamdamania87 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -31,22 +31,30 @@ std::shared_ptr<FutureMessage> py_rpc_builtin(
-        // ``createStackForSchema`` to avoid throwing an error.
-        Stack stack = torch::jit::createStackForSchema(
-            op->schema(), args, kwargs, c10::nullopt);
+    if (true) {
```
Forgot to push an update, this was something I was testing.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/dist_autograd_container.cpp:
```diff
@@ -10,8 +10,10 @@ constexpr int64_t kContextIdMask = (1LL << kContextIdBits) - 1;
-    : current_context_id_(0), worker_id_(0), initialized_(false) {}
+    : next_context_id_(0), worker_id_(0), initialized_(false) {}
 
 DistAutogradContainer& DistAutogradContainer::init(int64_t worker_id) {
```
Because we need to validate the worker_id below. If this is int16_t, we can pass a very large value (outside int16_t) and it will just wrap around.

pritamdamania87 commented on test/test_dist_autograd.py:
```diff
@@ -1,20 +1,25 @@
 
-import six
-import unittest
+import sys
```
I'm not sure.

pritamdamania87 commented:
> A code structure question: should dist_autograd be under the rpc folder? (if so, don't need to be addressed in this PR). I am asking because I was thinking about where we should put the worker_id_t type. This type should be shared by rpc and dist_autograd, and no other modules. But it does not make much sense to put this type directly under distributed folder.

Do we foresee any use case for dist_autograd without using rpc? If not in near future, shall we merge dist_autograd into rpc?

I think they should be in separate folders. For common stuff, we can put it under `distributed/common` and have both depend on them.

Commit:
Update on "Attach 'send' autograd function to the autograd graph as part of RPC."


This contains very basic functionality of adding 'send' autograd
function to our autograd graph. The purpose of this change is to validate the
basic structure proposed here makes sense. Once this makes sense, we can build
upon this to address more complicated scenarios. At a high level we've added
the following functionality:

1) Define a very simple 'SendRpcBackwards' autograd function.
2) Attach this function to appropriate tensors when we call an RPC.
3) Store the send function in our distributed autograd context.

GitHub Issue: https://github.com/pytorch/pytorch/issues/23110
Differential Revision: [D16903255](https://our.internmc.facebook.com/intern/diff/D16903255/)

pritamdamania87 commented:
@ezyang @mrshenli https://github.com/pytorch/pytorch/pull/25527, which is built on top of this will hopefully give you a much more comprehensive picture.

Commit:
Update on "Attach 'send' autograd function to the autograd graph as part of RPC."


This contains very basic functionality of adding 'send' autograd
function to our autograd graph. The purpose of this change is to validate the
basic structure proposed here makes sense. Once this makes sense, we can build
upon this to address more complicated scenarios. At a high level we've added
the following functionality:

1) Define a very simple 'SendRpcBackwards' autograd function.
2) Attach this function to appropriate tensors when we call an RPC.
3) Store the send function in our distributed autograd context.

GitHub Issue: https://github.com/pytorch/pytorch/issues/23110
Differential Revision: [D16903255](https://our.internmc.facebook.com/intern/diff/D16903255/)

Commit:
Update on "Attach 'send' autograd function to the autograd graph as part of RPC."


This contains very basic functionality of adding 'send' autograd
function to our autograd graph. The purpose of this change is to validate the
basic structure proposed here makes sense. Once this makes sense, we can build
upon this to address more complicated scenarios. At a high level we've added
the following functionality:

1) Define a very simple 'SendRpcBackwards' autograd function.
2) Attach this function to appropriate tensors when we call an RPC.
3) Store the send function in our distributed autograd context.

GitHub Issue: https://github.com/pytorch/pytorch/issues/23110
Differential Revision: [D16903255](https://our.internmc.facebook.com/intern/diff/D16903255/)
